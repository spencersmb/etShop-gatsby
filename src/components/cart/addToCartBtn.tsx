import { IState } from '@et/types/State'
import { addProductToCart, cartToggle, IAddProductAction } from '@redux/actions/cartActions'
import React from 'react'
import { Action, bindActionCreators, Dispatch } from 'redux'
import { connect } from 'react-redux'
import { ICartState } from '@et/types/Cart'
import { IProduct } from '@et/types/Products'

interface IPropsPublic {
	handleAddToCartState: () => void;
	isInCart: boolean;
	pwyw: null | number;
	slug: string;
	selectedProduct: IProduct;
	licenseQty: number | string;
	price: string
}

interface IPropsReduxActions {
	addToCart: IAddProductAction,
	cartToggle: () => void,
}

interface IPropsPrivate {
	cart: ICartState,
}

export class AddToCartBtn extends React.Component<IPropsPublic & IPropsPrivate & IPropsReduxActions> {
	static calcPriceBasedOnQty = (qty: number, price: string): string => {
		if (qty > 10) {
			const strToNum = parseInt(price, 10)
			const percent = .10
			return (strToNum - (percent * strToNum)).toString()
		} else {
			return price
		}
	}

	handleAddToCart = async () => {

		const { addToCart, cart, selectedProduct, handleAddToCartState, licenseQty, price } = this.props

		// add item, the current items in cart, the standard license slug, and qty
		if (typeof licenseQty !== 'string') {
			addToCart(selectedProduct, cart.items, this.props.slug, licenseQty, price)
			handleAddToCartState()
		}
	}

	checkout = () => {
		// Show cart pop-up
		this.props.cartToggle()
	}

	render () {
		// CART has been loaded variable to load in components? Or delay animation?
		const disabled = (this.props.licenseQty === 0) || (typeof this.props.licenseQty === 'string')

		if (this.props.isInCart) {
			return (
				<button onClick={this.checkout}>
					Checkout
				</button>
			)
		}

		return (
			<>
				<button onClick={this.handleAddToCart} disabled={disabled}>
					Add To Cart
				</button>
				{disabled && <div className='jestWarning'>Must have at least one computer license selected</div>}
			</>
		)
	}
}

const mapStateToProps = (state: IState): { cart: ICartState } => {
	return {
		cart: state.cart
	}
}

const mapDispatchToProps = (dispatch: Dispatch<Action>) => {
	return {
		addToCart: bindActionCreators(addProductToCart, dispatch),
		cartToggle: bindActionCreators(cartToggle, dispatch)
	}
}

export default connect<IPropsPrivate, IPropsReduxActions, IPropsPublic, IState>(mapStateToProps, mapDispatchToProps)(AddToCartBtn)
