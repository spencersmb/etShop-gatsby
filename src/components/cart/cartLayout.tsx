import CartList from '@components/cart/cartList'
import CheckoutTabs from '@components/tabs/checkoutTabs'
import { IProducts } from '@et/types/Products'
import { IState } from '@et/types/State'
import { cartToggle, changeCheckoutType, emptyCart } from '@redux/actions/cartActions'
import { isPWYWItemInCart } from '@utils/cartUtils'
import { connect } from 'react-redux'
import styled from 'styled-components'
import { ICartState } from '@et/types/Cart'
import { Action, bindActionCreators, Dispatch } from 'redux'
import { Actions } from '@et/types/Actions'
import React, { RefObject } from 'react'

interface IPropsPublic {
	poseRef: RefObject<any>;
}

interface IReduxState {
	cart: ICartState,
	products: IProducts,
}

interface IReduxActions {
	changeCheckout: (type: string) => Actions,
	emptyCart: () => Actions,
	cartToggle: () => void
}

function preventDefault (e) {
	e = e || window.event
	if (e.preventDefault) {
		e.preventDefault()
	}
	e.returnValue = false
}

export class CartLayout extends React.Component<IPropsPublic & IReduxState & IReduxActions> {

	getCheckOutForm = () => {
		const { cart } = this.props
		// TODO: Redo how we check for free checkout
		// Possibly add a free checkout toggle into redux so if the cart is empty and we
		// add a free item - the cart is toggled to free checkout
		// if cart has an item in it already and is not set to free, then the checkout is
		// not changed to free
		if (this.props.cart.totalPrice === 0 && isPWYWItemInCart(this.props.cart.items, this.props.products)) {
			return (<div className='freeCheckout'>Free item checkout</div>)
		} else {
			return (
				<CheckoutTabs
					cartTotal={this.props.cart.totalPrice}
					initialLoad='stripe'
					handleChangeType={this.props.changeCheckout}
				>
					<div data-payment='stripe'>Tab 1</div>
					<div data-payment='paypal'>Tab 2</div>
				</CheckoutTabs>
			)
		}

	}

	componentDidMount (): void {
		const body: HTMLElement | null = document.getElementById('app')
		if (body) {
			body.style.top = '-' + window.pageYOffset.toString(10) + 'px'
			body.style.position = 'fixed'
		}

	}

	componentWillUnmount (): void {
		const body = document.getElementById('app')
		if (body) {
			const styles = window.getComputedStyle(body, 'top')
			console.log('parseInt(styles.top.substring(1), 10)', parseInt(styles.top.substr(1), 10))

			body.style.position = 'relative'
			window.scrollTo(0, (styles.top ? parseInt(styles.top.substr(1), 10) : 0))
			body.style.top = 'auto'
		}

	}

	render () {
		return (
			<CartWrapper data-testid='cart-wrapper' ref={this.props.poseRef}>
				<button data-testid='close-btn' className='jestCloseCart' onClick={this.props.cartToggle}>Close</button>

				<div>
					<button
						data-testid='empty-cart-btn'
						className='jestEmptyCart'
						onClick={this.props.emptyCart}>
						Empty Cart
					</button>
				</div>

				<div>
					<CartList/>
				</div>

				<div>
					<CheckoutTabs
						cartTotal={this.props.cart.totalPrice}
						initialLoad='stripe'
						handleChangeType={this.props.changeCheckout}
						freeCheckout={this.props.cart.totalPrice === 0 && isPWYWItemInCart(this.props.cart.items, this.props.products)}
					>
						<div data-payment='stripe'>Tab 1</div>
						<div data-payment='paypal'>Tab 2</div>
					</CheckoutTabs>
					<hr/>
					<div>
						<div>pay what you want found?</div>
						<div>
							{JSON.stringify(isPWYWItemInCart(this.props.cart.items, this.props.products))}
						</div>
					</div>
					<hr/>
					<div>
						<button
							type='button'
							disabled={this.props.cart.totalPrice === 0
								? !isPWYWItemInCart(this.props.cart.items, this.props.products)
								: false}>Checkout temp btn
						</button>
					</div>
				</div>
			</CartWrapper>
		)
	}
}

const CartWrapper = styled.div`
	position: fixed;
	top: 0;
	left: 0;
	height: 100vh;
	width: 100%;
	overflow-y: scroll;
	background: #8ac3c0;
	transform:translateY(0) translateX(0);
	z-index: 4;
	//transition: all 300ms cubic-bezier(0.785, 0.135, 0.15, 0.86);
`

const mapStateToProps = (state: IState): any => {
	return {
		cart: state.cart,
		products: state.products
	}
}

const mapDispatchToProps = (dispatch: Dispatch<Action>) => {
	return {
		cartToggle: bindActionCreators(cartToggle, dispatch),
		changeCheckout: bindActionCreators(changeCheckoutType, dispatch),
		emptyCart: bindActionCreators(emptyCart, dispatch)
	}
}

export default connect<IReduxState, IReduxActions, IPropsPublic, IState>(mapStateToProps, mapDispatchToProps)(CartLayout)
