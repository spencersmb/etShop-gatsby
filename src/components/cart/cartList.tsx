import CartItem from '@components/cart/cartItem'
import { ICartState } from '@et/types/Cart'
import { IState } from '@et/types/State'
import React from 'react'
import { connect } from 'react-redux'

interface IProps {
	cart: ICartState,
}

export function CartList (props: IProps) {
	const { cart } = props
	const itemsKeyArray: string[] = Object.keys(cart.items)

	if (itemsKeyArray.length < 1) {
		return (
			<div>Your cart is empty</div>
		)
	}
	return (
		<div data-testid='cartList'>
			{itemsKeyArray
				.map((slug, index) =>
					<CartItem
						cartIndex={slug}
						key={index}
					/>
				)}
		</div>
	)
}

const mapStateToProps = (state: IState): any => {
	return {
		cart: state.cart
	}
}

export default connect<IProps, {}, {}, IState>(mapStateToProps)(CartList)