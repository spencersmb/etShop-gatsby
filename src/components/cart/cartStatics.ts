export class CartPricingConfig {
	static minQuantity = 10
	static bulkDiscount = .1
	static discountTier1 = .1
	static discountTier2 = .2
	static discountTier3 = .3
}
