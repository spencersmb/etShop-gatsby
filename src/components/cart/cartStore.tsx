import PoseHoc, { IPoseHoc } from '@components/animations/poseHoc'
import { OnPoseComplete } from '@et/types/Modal'
import { IState } from '@et/types/State'
import { cartLoadedComplete as cartLoaded, cartToggle, updateCartState } from '@redux/actions/cartActions'
import { getLocalStorageCart } from '@utils/cartUtils'
import React, { useLayoutEffect, useRef } from 'react'
import { connect } from 'react-redux'
import { Action, bindActionCreators, Dispatch } from 'redux'
import styled from 'styled-components'
import posed, { PoseGroup } from 'react-pose'
import { ILocalStorageCart } from '@et/types/Cart'
import Cart from '@components/cart/cartLayout'

interface IPropsPrivate {
	cartIsOpen: boolean;
}

interface IPrivateActions {
	cartLoadedComplete: () => any;
	updateCartState: (cart: ILocalStorageCart) => any,
	openCart: () => void
}

interface IPropsPublic {
	defaultOpenState?: boolean
}

export const MyShoppingCart = (props: IPropsPrivate & IPrivateActions & IPropsPublic) => {
	const target = useRef<HTMLElement | null>(null)
	const bodyScrollPos = useRef(0)
	const { cartIsOpen } = props
	// onComponent mount
	useLayoutEffect(() => {
		target.current = document.querySelector('#___gatsby')

		const localStateCart: ILocalStorageCart = getLocalStorageCart()

		if (Object.keys(localStateCart).length > 0) {
			// update Cart with data from local storage
			props.updateCartState(localStateCart)
		}
		props.cartLoadedComplete()

		if (props.defaultOpenState) {
			props.openCart()
		}

	}, [])

	return (
		<CartStyled id='cart-Container' style={{ position: 'relative', zIndex: 4 }}>
			<PoseGroup>
				{cartIsOpen &&
        <CartPose key='cart' onPoseComplete={(type: OnPoseComplete) => {

					const overlayOpen = document.querySelector('#overlay')
					if (type === 'enter' && !overlayOpen && target.current) {
						// MOVED TO NAV so happen faster
						bodyScrollPos.current = document.body.scrollTop || document.documentElement.scrollTop || 0
						target.current.style.width = `100%`
						target.current.style.top = `-${bodyScrollPos.current}px`
						target.current.style.bottom = `0`
						target.current.style.position = 'fixed'
					}

					if (type === 'exit' && !overlayOpen && target.current) {
						target.current.style.removeProperty('position')
						target.current.style.removeProperty('top')
						target.current.style.removeProperty('bottom')
						target.current.style.removeProperty('padding')

						document.documentElement.scrollTop = document.body.scrollTop = bodyScrollPos.current
					}

				}}>
					{({ ref }: IPoseHoc) => (
						<Cart poseRef={ref}/>
					)}
        </CartPose>
				}
			</PoseGroup>
		</CartStyled>
	)
}

const mapStateToProps = (state: IState): { cartIsOpen: boolean } => {
	return {
		cartIsOpen: state.cart.isOpen
	}
}
const mapDispatchToProps = (dispatch: Dispatch<Action>): IPrivateActions => {
	return {
		cartLoadedComplete: bindActionCreators(cartLoaded, dispatch),
		openCart: bindActionCreators(cartToggle, dispatch),
		updateCartState: bindActionCreators(updateCartState, dispatch)
	}
}

export default connect<IPropsPrivate, IPrivateActions, IPropsPublic, IState>(mapStateToProps, mapDispatchToProps)(MyShoppingCart)

// Styles
const CartStyled = styled.div`
	
`

// animations
const CartPose = posed(PoseHoc)({
	enter: {
		opacity: 1,
		transition: {
			default: {
				ease: 'easeOut'
			}
		}
	},
	exit: {
		opacity: 0,
		transition: {
			default: {}
		}
	}
})
