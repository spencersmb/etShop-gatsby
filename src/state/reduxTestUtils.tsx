// this is a handy function that I normally make available for all my tests
import { ICartState } from '@et/types/Cart'
import { IGatsbyConfig } from '@et/types/Gatsby'
import { IProduct } from '@et/types/Products'
import React from 'react'
import initialState from '@redux/reducers/initialState'
import { Provider } from 'react-redux'
import { render } from 'react-testing-library'
// that deal with connected components.
// you can provide initialState or the entire store that the ui is rendered with
import { createStore } from 'redux'

export function renderWithRedux (
	ui: any,
	reducer: any,
	{ state, store = createStore(reducer, state) }: any = initialState
) {
	return {
		...render(<Provider store={store}>{ui}</Provider>),
		// adding `store` to the returned utilities to allow us
		// to reference it in our tests (just try to avoid using
		// this to test implementation details).
		store
	}
}

const testSiteMetaDataQuery = {
	title: `Every-Tuesday Digital Products Shop`,
	description: `Kick off your next, great Gatsby project with this default starter. This barebones starter ships with the main Gatsby configuration files you might need.`,
	author: `@Teelac`,
	authorUrl: 'https://every-tuesday.com/about/#teela',
	siteUrl: `https://shop.every-tuesday.com`,
	siteName: `Every-Tuesday Shop`,
	db: 'http://shopeverytuesday.local',
	route: 'et-shop',
	twitterUrl: 'https://twitter.com/teelacunningham',
	twitterDefaultImage: 'url'
}

export const singleItemQuery: IGatsbyConfig & { wcProduct: IProduct } = {
	site: {
		siteMetadata: {
			...testSiteMetaDataQuery
		}
	},
	wcProduct: {
		type: 'simple',
		date_created_gmt: '2018-09-25T20:35:51',
		date_modified_gmt: '2019-02-18T21:44:25',
		id: '202eca74-fc90-56e7-8269-b59f18a19194',
		name: 'Watercolor texture kit Vol. 1',
		price: '16',
		product_id: 222,
		pwyw: false,
		regular_price: '16',
		sale_price: '',
		slug: 'watercolor-texture-kit-vol-1',
		description: 'description',
		short_description: 'short desc',
		on_sale: false,
		features: {
			description: 'feature desc',
			items: {
				description: 'item desc',
				icon: 'icon',
				title: 'singleItemQuery title'
			}
		},
		categories: [
			{ id: 1, slug: 'fonts', name: 'Fonts' },
			{ id: 2, slug: 'Watercolor', name: 'Watercolor' }
		],
		images: [
			{
				id: 123414,
				alt: 'alt',
				fullSize: {
					url: 'fullsize-url'
				},
				thumbnail: {
					url: 'image-url'
				}
			}
		],
		license: {
			hasExtendedLicense: true,
			type: 'standard',
			extendedItem: {
				slug: 'watercolor-texture-kit-vol-1-ext'
			}
		},
		seo: {
			desc: 'Add serious vibrant color and detail with this kit packed with 32 unique watercolor textures, 12 seamless, repeatable watercolor texture patterns and 2 bonus watercolor paper patterns.',
			title: 'Watercolor texture kit Vol. 1'
		},
		tags: [
			{ id: 1, name: 'Texture Kit', slug: 'texture-kit' },
			{ id: 2, name: 'Waterbrush', slug: 'waterbrush' }
		],
		localFile: {
			name: 'file name',
			id: '1234',
			childImageSharp: {
				fluid: {
					src: 'src'
				},
				fixed: {
					width: '400',
					height: '400',
					src: 'src'
				}
			}
		}
	}
}

export const testCartEmpty: ICartState = {
	couponCode: {
		code: 'string',
		discount: 'string',
		loading: false,
		product_ids: [],
		submitted: false,
		type: 'string',
		valid: false,
	},
	isOpen: false,
	items: {
	},
	loaded: true,
	paymentType: 'Stripe',
	totalItems: 0,
	totalPrice: 0
}
