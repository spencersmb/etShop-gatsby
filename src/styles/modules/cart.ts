import { device } from '@styles/global/breakpoints'
import { colors } from '@styles/global/colors'
import { col, GridFluid } from '@styles/global/cssGrid'
import { Sentinel } from '@styles/global/fonts'
import styled from 'styled-components'

export const CartPageContainer = styled.div`
	width: 100%;
	padding: 20px;
	position: relative;
	@media ${device.tablet} {
	}
	@media ${device.laptop} {
		display: grid;
		grid-template-columns:
			minmax(0, 1fr)
			repeat(12, minmax(30px, ${col}))
			minmax(0, 1fr)
		;
		grid-gap: 30px;
		align-items: flex-start;
		padding: 20px 0 0;
	}
`
export const CartSliderTransition = {
	enter: {
		duration: 600,
		ease: [1, 0, 0, 1]
	},
	exit: {
		ease: [1, 0, 0, 1]
	}
}
export const CartHeader = styled.div`
	border-bottom: 1px solid ${colors.grey.i300};
	padding: 15px;
	grid-column: 1 / -1;
	
	.cartHeader__inner{
		display: flex;
		flex-direction: row;
		justify-content: center;
	}

	.closeCartBtn,
	.closeCheckoutBtn{
		width: 56px;
		height: 60px;
		display: flex;
		svg{
		max-width: 36px;
			width: 100%;
		}
		path{
			fill: ${colors.grey.i800};
		}
		
		@media ${device.tablet} {
			position: absolute;
			left: 20px;		    
		}
		
		@media ${device.laptop} {
			&:hover{
				cursor: pointer;
			}
		}
			
	}
	
	.emptyCartBtn{
		width: 56px;
	}
`
export const CartHeaderTitle = styled.div`
	
	font-weight: bold;
	text-align: center;
	flex: 1;
	display: flex;
	justify-content: center;
	align-items: center;
	flex-direction: column;

	
	h2{
		${Sentinel.black};
		line-height: 28px;
		font-size: 28px;
		color: ${colors.grey.i800};
	}
	p{
		${Sentinel.reg};
		margin: 0;
		color: ${colors.secondary.text};
	}
	.total{
		font-size: 24px;
		line-height: 24px;
		color: ${colors.grey.i800};
	}
	.qty{
		display: none;
	}
	@media ${device.tablet} {
		.qty{
			display: flex;
			flex-direction: column;
		}
		.total{
			display: none;
		}
		
	}
		
`
export const EmptyCartList = styled.div`
	text-align: center;
	width: 100%;
	font-size: 18px;
	${Sentinel.semiboldItalic};
	margin-top: 25px;
	color: ${colors.grey.i800};
`
