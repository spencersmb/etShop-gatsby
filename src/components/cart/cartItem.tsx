import { CartPricingConfig } from '@components/cart/cartStatics'
import NumberDial from '@components/forms/inputs/numberDial'
import { IState } from '@et/types/State'
import {
	removeProductFromCart,
	updateCartItemQty as updateCartItem
} from '@redux/actions/cartActions'
import { colors } from '@styles/global/colors'
import {
	CartItemContainer,
	CartItemDetails,
	CartItemLicense,
	CartItemDetail,
	CartItemTitle,
	CartItemDiscount,
	VolumeDiscountPin,
	CartItemHeader,
	CartItemContent,
	CartItemBorder, CartItemDashedBorder, RemoveItemMobile, RemoveItemDesktop
} from '@styles/modules/cartItem'
import { svgs } from '@svg'
import { checkForCoupon } from '@utils/cartUtils'
import { calcCouponDiscount, displayCurrency, displayPercent, getPrice } from '@utils/priceUtils'
import { renderSvg } from '@utils/styleUtils'
import React, { useEffect, useRef, useState } from 'react'
import { IProduct, IProducts } from '@et/types/Products'
import { ICartState, IChangeLicenseData, IChangeQty } from '@et/types/Cart'
import { Action, bindActionCreators, Dispatch } from 'redux'
import { connect } from 'react-redux'

// import styled from 'styled-components'

/**
 * CartItem properties.
 */
interface IProps {
	cartIndex: string;
	itemIndex: number
}

interface IReduxProps {
	cart: ICartState;
	products: IProducts;
}

interface IReduxPropActions {
	updateCartItemQty: (data: IChangeQty) => void;
	changeLicense: (data: IChangeLicenseData) => void;
	removeItem: (slug: string) => void;
}

/**
 * CartItem.
 *
 * How it works:
 * The cart item controls what license is selected. It determines if a license is standard or extended based off of
 * whether the item in the cart has a property "extended" as either true or false.
 *
 * To change the value, we look at what value was selected by the user and pass the object to our action to change it in redux.
 * The change is then reflected in the UI.
 * {
 * 	itemSlug: this.props.cartIndex, // primary Slug name to look up the same item
 * 	extended: isExtended, // which license type are we trying to select?
 * 	products: this.props.products, // Full list of products to ref and get the new license + pricing data for
 * 	currentCartItem: this.selectedCartItem // current cart item to get the current qty selected
 * }
 *
 * Item total calculates the coupon discount on the fly so when a user saves the cart we dont save the discount
 * Same thing is done for the total cart, we check for the coupon on the fly when creating the total
 */
export function CartItem (props: IProps & IReduxProps & IReduxPropActions) {
	const { cart, cartIndex, products, removeItem } = props
	const [bulkDiscount, setBulkDiscount] = useState(false)
	const total = parseInt(cart.items[cartIndex].price, 10) * convertToNumber(cart.items[cartIndex].qty)
	const cartItem = cart.items[cartIndex]

	useEffect(() => {
		if (cart.items[cartIndex].qty > CartPricingConfig.minQuantity) {
			setBulkDiscount(true)
		}
	}, [])

	function convertToNumber (item: any): number {
		if (typeof item === 'string') {
			return parseInt(item, 10)
		} else {
			return item
		}
	}

	function hasCoupon (): boolean {
		return checkForCoupon(cart.coupon.product_ids, cart.items[cartIndex].id)
	}

	function calcDiscount (): string {
		const price = parseInt(cart.items[cartIndex].price, 10)
		return displayCurrency(calcCouponDiscount(cart.coupon, price))
	}

	function handleRemoveItem () {
		removeItem(cartIndex)
	}

	function showLicenseType () {
		if (cart.items[cartIndex].extended) {
			return (
				<CartItemLicense
					type={'extended'}>
					Extended License
				</CartItemLicense>
			)
		}

		return (
			<CartItemLicense
				type={'standard'}>
				Standard License
			</CartItemLicense>
		)
	}

	function displaySavings () {
		const originalTotal = parseInt(products[cartItem.slug].price, 10) * cartItem.qty
		const currentTotal = parseInt(cartItem.price, 10) * cartItem.qty
		return displayCurrency(originalTotal - currentTotal)
	}

	function displayOriginalTotal () {
		return displayCurrency(parseInt(products[cartItem.slug].price, 10) * cartItem.qty)
	}

	return (
		<CartItemContainer>
			<CartItemBorder className={props.itemIndex === 0 ? 'top' : ''}>
				<svg viewBox='0 0 709 10' fill='none' xmlns='http://www.w3.org/2000/svg'>
					<path fillRule='evenodd' clipRule='evenodd'
								d='M331 0C331 4.41827 327.418 8 323 8C318.582 8 315 4.41827 315 0H15C8.46889 0 2.91269 4.17401 0.853516 10H708.146C706.087 4.17401 700.531 0 694 0H331Z'
								fill='#525252'/>
				</svg>

			</CartItemBorder>

			<CartItemContent>
				<CartItemHeader>
					<CartItemTitle data-testid='productName'>
						{products[cartIndex].name}
					</CartItemTitle>

					{showLicenseType()}

					<RemoveItemDesktop>
						<button onClick={handleRemoveItem} data-testid='removeItemBtn'>
							Remove Item
						</button>
					</RemoveItemDesktop>
				</CartItemHeader>

				<CartItemDashedBorder>
					<svg viewBox='0 0 2 186' fill='none' xmlns='http://www.w3.org/2000/svg'>
						<path d='M1 0L0.999992 186' stroke='#DADADA' strokeDasharray='5 5'/>
					</svg>

				</CartItemDashedBorder>

				<CartItemDetails>

					{/*Price Per License*/}
					<CartItemDetail>
						<span>Price per license</span>
						<p data-testid='productPrice'>{displayCurrency(cart.items[cartIndex].price)}</p>
					</CartItemDetail>

					{/*License Qty*/}
					<CartItemDetail>
						<span>License Qty</span>
						<p>{cart.items[cartIndex].qty}</p>
					</CartItemDetail>

					{/*Discount*/}
					{bulkDiscount &&
          <>
            <CartItemDiscount>
              <span>Original Total</span>
              <p>{displayOriginalTotal()}</p>
            </CartItemDiscount>
            <CartItemDiscount>
              <span className={'discountLabel'}>{displayPercent(CartPricingConfig.bulkDiscount)}% Savings</span>
              <div className={'discountPin'}>
                <VolumeDiscountPin>
                  Volume Discount
                </VolumeDiscountPin>
              </div>
              <div className={'discount'}>-{displaySavings()}</div>
            </CartItemDiscount>
          </>
					}

					{/*Total*/}
					<CartItemDetail total={true}>
						<span className={'totalLabel'}>total</span>
						<div className={'total'}>{displayCurrency(total)}</div>
					</CartItemDetail>
				</CartItemDetails>

				<RemoveItemMobile>
					<button onClick={handleRemoveItem} data-testid='removeItemBtn'>
						Remove Item
					</button>
				</RemoveItemMobile>
			</CartItemContent>

			<CartItemBorder bottom={true}>
				<svg viewBox='0 0 709 10' fill='none' xmlns='http://www.w3.org/2000/svg'>
					<path fillRule='evenodd' clipRule='evenodd'
								d='M0.853516 0C2.91269 5.82599 8.46889 10 15 10H315C315 5.58173 318.582 2 323 2C327.418 2 331 5.58173 331 10H694C700.531 10 706.087 5.82599 708.146 0H0.853516Z'
								fill='#525252'/>
				</svg>
			</CartItemBorder>
		</CartItemContainer>
	)
}

const mapStateToProps = (state: IState): { cart: ICartState, products: IProducts } => {
	return {
		cart: state.cart,
		products: state.products
	}
}

const mapDispatchToProps = (dispatch: Dispatch<Action>): any => {
	return {
		removeItem: bindActionCreators(removeProductFromCart, dispatch),
		updateCartItemQty: bindActionCreators(updateCartItem, dispatch)
	}
}

export default connect<IReduxProps, IReduxPropActions, IProps, IState>(mapStateToProps, mapDispatchToProps)(CartItem)

