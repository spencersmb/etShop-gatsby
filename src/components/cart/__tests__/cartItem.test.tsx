import { CartItem } from '@components/cart/cartItem'
import { ProductKey, testCartWithItem, testProducts } from '@redux/reduxTestUtils'
import { calcBulkDiscount, displayCurrency, getPrice } from '@utils/priceUtils'
import React from 'react'
import renderer from 'react-test-renderer'
import {
	render,
	cleanup,
	fireEvent
} from 'react-testing-library'
import 'jest-dom/extend-expect'

afterEach(cleanup)

const props = {
	cartIndex: ProductKey.WatercolorStd,
	cart: {
		...testCartWithItem
	},
	products: testProducts,
	updateCartItemQty: jest.fn(),
	changeLicense: jest.fn(),
	removeItem: jest.fn()
}

describe('User Hoc Render Props', () => {

	// TODO: add coupon & discount tests
	it('renders correctly', () => {
		const tree = renderer
			.create(
				<CartItem {...props}/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('Should render correct item name', () => {
		const modalRender = render(<CartItem {...props}/>)
		expect(modalRender.getByTestId('productName').innerHTML).toBe(testProducts[ProductKey.WatercolorStd].name)
	})

	it('Should render correct regular price', () => {
		const modalRender = render(<CartItem {...props}/>)
		expect(modalRender.getByTestId('productPrice').innerHTML).toBe(displayCurrency(testProducts[ProductKey.WatercolorStd].price))
	})

	it('Should render correct license', () => {
		const modalRender = render(<CartItem {...props}/>)
		const select = modalRender.getByTestId('selectID')
		expect(select.children[0].innerHTML).toEqual('Standard')
	})

	it('Should render correct discounted price and show discount content', () => {
		const discountedProps = props
		discountedProps.cart.items[ProductKey.WatercolorStd].qty = 12
		discountedProps.cart.items[ProductKey.WatercolorStd].price = '14.4'
		const modalRender = render(<CartItem {...discountedProps}/>)
		const input: any = modalRender.getByTestId('numberDial')

		expect(modalRender.getByTestId('bulkDiscount').innerHTML).toEqual('Bulk discount of 0.1 applied')
		expect(modalRender.getByTestId('productPrice').innerHTML).toBe(displayCurrency(calcBulkDiscount(testProducts[ProductKey.WatercolorStd].price)))
		expect(input.value).toEqual('12')
	})

	it('Should call updateCartItemQty action', () => {
		const modalRender = render(<CartItem {...props}/>)
		const input: any = modalRender.getByTestId('numberDial')
		fireEvent.input(input, { target: { value: 23 } })

		const cartItemRef = props.cart.items[ProductKey.WatercolorStd]
		const calledWidth = {
			key: props.cartIndex,
			cartItem: {
				...cartItemRef,
				qty: 23
			},
			bulkDiscount: true,
			regularPrice: getPrice(testProducts[cartItemRef.slug])
		}
		expect(props.updateCartItemQty).toHaveBeenCalledTimes(1)
		expect(props.updateCartItemQty).toHaveBeenCalledWith(calledWidth)

	})

	it('Should call changeLicense action', () => {
		const modalRender = render(<CartItem {...props}/>)
		const select = modalRender.getByTestId('selectID')
		fireEvent.change(select)

		const calledWith = {
			currentCartItem: props.cart.items[props.cartIndex], // item in cart
			extended: false, // value from dropdown
			cartItemIndex: props.cartIndex, // redundent?
			products: testProducts, // all our products
			bulkDiscount: true // is bulkdiscount enabled
		}
		expect(props.changeLicense).toHaveBeenCalledTimes(1)
		expect(props.changeLicense).toHaveBeenCalledWith(calledWith)

	})

	it('Should call removeItem action with correct item index', () => {
		const modalRender = render(<CartItem {...props}/>)
		const btn = modalRender.getByTestId('removeItemBtn')
		btn.click()
		expect(props.removeItem).toHaveBeenCalledTimes(1)
		expect(props.removeItem).toHaveBeenCalledWith(ProductKey.WatercolorStd)
	})

})