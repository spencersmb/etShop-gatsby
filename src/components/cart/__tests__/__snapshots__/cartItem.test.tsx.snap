// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Hoc Render Props renders correctly 1`] = `
<div>
  <div
    data-testid="productName"
  >
    Watercolor texture kit Vol. 1
  </div>
  <div>
    <select
      data-testid="selectID"
      onChange={[Function]}
      value="standard"
    >
      <option
        value="standard"
      >
        Standard
      </option>
      <option
        value="extended"
      >
        Extended
      </option>
    </select>
  </div>
  <div>
    <label
      htmlFor="numberDial"
    >
      Quantity
    </label>
    <input
      className="numberInput"
      data-testid="numberDial"
      id="numberDial"
      onChange={[Function]}
      onKeyPress={[Function]}
      readOnly={false}
      type="number"
      value={1}
    />
  </div>
  <div>
    Item:
  </div>
  <div
    data-testid="productPrice"
  >
    $16.00
  </div>
  <button
    data-testid="removeItemBtn"
    onClick={[Function]}
  >
    Remove Item
  </button>
  <hr />
</div>
`;
